### PR TITLE
Add CapacityBasedDeterministicHashPolicy

### PR DESCRIPTION
**Pull request by dbw9580**
### What changes are proposed in this pull request?

Add a new block location policy `CapacityBaseDeterministicHashPolicy`.

### Why are the changes needed?

We want a `CapacityBaseRandomPolicy` that is deterministic.

See also #16187.

### Does this PR introduce any user facing changes?

Yes, a new block location policy is available for config item `alluxio.user.ufs.block.read.location.policy` and `alluxio.user.block.write.location.policy.class`.

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseHashPolicy.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * so that workers with more capacity have a higher chance of being chosen.
+ * The difference between this policy and {@link CapacityBaseRandomPolicy} is that this policy
+ * uses the hashed block ID as the index to choose the target worker, so that the same block is
+ * always routed to the same worker.
+ * {@link MurmurHash3} is used as the hashing function to simulate a uniformly distributed random
+ * source.
+ *
+ * @see  CapacityBaseRandomPolicy
+ */
+public class CapacityBaseHashPolicy implements BlockLocationPolicy {
```
[the remark this answers is not in the document]
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseHashPolicy.java:35]: +1

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseHashPolicy.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+   * Constructor required by
+   * {@link BlockLocationPolicy.Factory#create(Class, AlluxioConfiguration)}.
+   * @param conf Alluxio configuration
+   */
+  public CapacityBaseHashPolicy(AlluxioConfiguration conf) {}
+
+  @Override
+  public Optional<WorkerNetAddress> getWorker(GetWorkerOptions options) {
+    // cumulative distribution function of worker capacities
+    TreeMap<Long, BlockWorkerInfo> capacityCdf = new TreeMap<>();
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseHashPolicy.java:46]: what does `cdf` mean?   We recommend using less abbreviations in variable names

Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseHashPolicy.java]: cumulative distribution function, I put the full name in the comment above the variable. It's a really long name.

**Comment by dbw9580** [comment]
https://github.com/Alluxio/alluxio/blob/da543a5b604cae9703d55dad4102996ab685c3b1/core/client/fs/src/main/java/alluxio/client/block/policy/DeterministicHashPolicy.java#L62-L66

`DeterministicHashPolicy` supports a num of shard config. I'm not sure if `CapacityBasedDeterministicHashPolicy` should do the same. 

@maobaolong @dengweisysu could you please shed some light on your cases, e.g. how many number of shards you use for `DeterministicHashPolicy`, and if `CapacityBasedDeterministicHashPolicy` should support more than one shard (the current impl is equivalent to 1 shard)?

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseHashPolicy.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseHashPolicy.java]: Do we want to add an alias `CapacityBasedHashPolicy` for `CapacityBaseRandomPolicy`? The inconsistency is annoying...

**Comment by dengweisysu** [comment]
> https://github.com/Alluxio/alluxio/blob/da543a5b604cae9703d55dad4102996ab685c3b1/core/client/fs/src/main/java/alluxio/client/block/policy/DeterministicHashPolicy.java#L62-L66
> 
> `DeterministicHashPolicy` supports a num of shard config. I'm not sure if `CapacityBasedDeterministicHashPolicy` should do the same.
> 
> @maobaolong @dengweisysu could you please shed some light on your cases, e.g. how many number of shards you use for `DeterministicHashPolicy`, and if `CapacityBasedDeterministicHashPolicy` should support more than one shard (the current impl is equivalent to 1 shard)?

@dbw9580 @maobaolong I suggest to support multi shards in CapacityBasedDeterministicHashPolicy, because one shard may result in IO bottleneck in highly concurrency. And background job can't ensure consistent  distribution  of replication with this policy forever.

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBasedDeterministicHashPolicy.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    //                     ^ start = 95
+    // 3. find the corresponding worker in the CDF.
+    //    which is worker B in this example
+    // 4. if #shards = 1, this worker is selected. otherwise, find a set of candidates:
+    //    4.1 hashed_block_id(0) = block id
+    //    4.2 for i in [1, #shards], hashed_block_id(i) = hash(hashed_block_id(i-1))
+    //    4.3 find the worker whose position corresponds to hashed_block_id(i) in the CDF,
+    //        and add it to the candidates set
+    //    4.4 repeat 4.2 - 4.4
+    // 5. select a random worker in the candidate set
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBasedDeterministicHashPolicy.java:87]: nit: put the these algorithm description to the class level java doc?  your call

**Review by beinan — APPROVED** [review]
lgtm

File: core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBasedDeterministicHashPolicy.java (hunk 3 above)
Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBasedDeterministicHashPolicy.java]: Done

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**